controller/volume_attachment_controller.go

```diff
@@ -0,0 +1,529 @@
+		return err
+	}
+	if namespace != vac.namespace {
+		return nil
+	}
+	return vac.reconcile(name)
+}
+
+func (vac *VolumeAttachmentController) reconcile(vaName string) (err error) {
+	va, err := vac.ds.GetLHVolumeAttachment(vaName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	isResponsible, err := vac.isResponsibleFor(va)
+	if err != nil {
+		return err
+	}
+	if !isResponsible {
+		return nil
+	}
+
+	//log := getLoggerForLHVolumeAttachment(vac.logger, va)
+
+	vol, err := vac.ds.GetVolume(va.Spec.Volume)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			if !va.DeletionTimestamp.IsZero() {
+				return vac.ds.RemoveFinalizerForLHVolumeAttachment(va)
+			}
+			return nil
+		}
+		return err
+	}
+
+	existingVA := va.DeepCopy()
+	existingVol := vol.DeepCopy()
+	defer func() {
+		if err != nil {
+			return
+		}
+		if !reflect.DeepEqual(existingVol.Spec, vol.Spec) {
+			if _, err = vac.ds.UpdateVolume(vol); err != nil {
+				return
+			}
+
+		}
+		if !reflect.DeepEqual(existingVA.Status, va.Status) {
+			if _, err = vac.ds.UpdateLHVolumeAttachmentStatus(va); err != nil {
+				return
+			}
+		}
+		return
+	}()
+
+	// Note that in this controller the desire state is recorded in VA.Spec
+	// and the current state of the world is recorded inside volume CR
+
+	vac.handleVolumeDetachment(va, vol)
+
+	vac.handleVolumeAttachment(va, vol)
+
+	return vac.handleVAStatusUpdate(va, vol)
+}
+
+func (vac *VolumeAttachmentController) handleVolumeDetachment(va *longhorn.VolumeAttachment, vol *longhorn.Volume) {
+	// Volume is already trying to detach
+	if vol.Spec.NodeID == "" {
+		return
+	}
+
+	if !shouldDoDetach(va, vol) {
+		return
+	}
+
+	// There is no attachmentSpec that request the current vol.Spec.NodeID.
+	// Therefore, set desire state of volume to empty
+	vol.Spec.NodeID = ""
+	// reset the attachment parameter for vol
+	setAttachmentParameter(map[string]string{}, vol)
+	return
+}
+
+func shouldDoDetach(va *longhorn.VolumeAttachment, vol *longhorn.Volume) bool {
+	// For auto salvage logic
+	// TODO: create Auto Salvage controller to handle this logic instead of AD controller
+	if vol.Status.Robustness == longhorn.VolumeRobustnessFaulted {
+		return true
+	}
+	for _, attachmentSpec := range va.Spec.AttachmentSpecs {
+		// TODO: refactor this! Consider handle ignore the csi-attacher ticket for RWX volume
+		if vol.Spec.AccessMode == longhorn.AccessModeReadWriteMany &&
+			!vol.Spec.Migratable &&
+			attachmentSpec.Type == longhorn.AttacherTypeCSIAttacher {
+			continue
+		}
+		// Found one attachmentSpec that is still requesting volume to attach to the current node
+		if attachmentSpec.NodeID == vol.Spec.NodeID && verifyAttachmentParameters(attachmentSpec.Parameters, vol) {
+			return false
+		}
+	}
+	return true
+}
+
+func (vac *VolumeAttachmentController) handleVolumeAttachment(va *longhorn.VolumeAttachment, vol *longhorn.Volume) {
+	// Wait for volume to be fully detached
+	if vol.Spec.NodeID != "" ||
```

Comment:
Shoud we check if `v.status.CurrentNodeID` is `""`?

Reply:
I think we should check for spec so that we don't overwrite it. We wait for the handleVolumeDetachment function to clear it first. What do you think?